File: dclib/core/src/dclib/geometry/VectorUtils.java
package dclib.geometry;

import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Vector2;

public final class VectorUtils {
	
	/**
	 * Buffer used for floating point calculations involving vectors.
	 * Vector calculations don't always round well to exact numbers, so this buffer can be used to check if any 
	 * attribute of the vector (e.g. x, y, length) is near the expected number.
	 */
	public static final float BUFFER = 1e-5f;
	
	private VectorUtils() {
	}
	
	public static final Vector2 unit(final Vector2 vector) {
		float length = vector.len();
		return vector.cpy().scl(1 / length);
	}
	
	public static final Vector2 lengthened(final Vector2 vector, final float length) {
		return vector.cpy().scl(length / vector.len());
	}
	
	public static final Vector2 offset(final Vector2 from, final Vector2 to) {
		return to.cpy().sub(from);
	}

	public static final float angleToPoint(final Vector2 from, final Vector2 to) {
		Vector2 offset = offset(from, to);
		return MathUtils.atan2(offset.y, offset.x);
	}
	
	public static final Vector2 fromAngle(final float degrees) {
		return fromAngle(degrees, 1);
	}
	
	public static final Vector2 fromAngle(final float degrees, final float length) {
		Vector2 vector = new Vector2(length, 0);
		return vector.rotate(degrees);
	}
	
	public static final float relativeMiddle(final float pivotMiddle, final float objectLength) {
		return pivotMiddle - objectLength / 2;
	}
	
}

File: dclib/core/src/main/java/dclib/geometry/VectorUtils.kt
package dclib.geometry

import com.badlogic.gdx.math.Vector2

object VectorUtils {
    fun inv(vector: Vector2): Vector2 {
        return Vector2(1 / vector.x, 1 / vector.y)
    }

    fun sign(vector: Vector2): Vector2 {
        return Vector2(Math.signum(vector.x), Math.signum(vector.y))
    }

    fun unit(vector: Vector2): Vector2 {
        val length = vector.len()
        return vector.cpy().scl(1 / length)
    }

    fun offset(from: Vector2, to: Vector2): Vector2 {
        return to.cpy().sub(from)
    }

    /**
     * Returns the x component of a line between from and to, given the y value.
     * @param p1
     * *
     * @param p2
     * *
     * @param lineY absolute y of point along line
     * *
     * @return
     */
    fun getLineX(p1: Vector2, p2: Vector2, lineY: Float): Float {
        val offset = offset(p1, p2)
        val slope = offset.y / offset.x
        return p1.x + (lineY - p1.y) / slope
    }

    fun getLineY(p1: Vector2, p2: Vector2, lineX: Float): Float {
        val offset = offset(p1, p2)
        val slope = offset.y / offset.x
        return p1.y + (lineX - p1.x) * slope
    }

    fun toVector2(degrees: Float, length: Float): Vector2 {
        return Vector2(length, 0f).setAngle(degrees)
    }
}
